### PR TITLE
add function to escape charaters for XML

### LINKED ISSUE #17250: A function to escape charaters for HTML and XML.
`encodeXMLComponent`
`decodeXMLComponent`

(similar to `encodeURLComponent`, `decodeURLComponent`)

Should replace characters like `<`, `&` and `"`, `'` to the corresponding entities.

Question:
- is there a difference between HTML and XML in the set of characters needed to be encoded?
- is it legal to replace zero byte to `&#0;`? If not we will keep it as is.

Additional info:
- we will assume UTF-8 but not check the validity of sequences (leave invalid UTF-8 as is);

### PR DESCRIPTION
**Pull request by nautaa**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add function `encodeXMLComponent` to escape characters to place string into XML text node or attribute.

### CODE REVIEW

File: src/IO/WriteHelpers.h [hunk 1]
```diff
@@ -581,6 +581,59 @@ void writeCSVString(const StringRef & s, WriteBuffer & buf)
     writeCSVString<quote>(s.data, s.data + s.size, buf);
 }
 
+inline void writeHTMLString(const char * begin, const char * end, WriteBuffer & buf)
+{
+    const char * pos = begin;
+    while (true)
+    {
+        const char * next_pos = find_first_symbols<'<', '&', '>', '"', '\''>(pos, end);
```
Comment (alexey-milovidov) [inline src/IO/WriteHelpers.h:589]: BTW, `>` is not necessary to escape.

Comment (alexey-milovidov) [inline src/IO/WriteHelpers.h:589]: What about control characters? Should we replace \0 to `&#0;` or it doesn't make sense?

Comment (alexey-milovidov) [reply src/IO/WriteHelpers.h:589]: Ok, let's keep it just in case.

File: src/IO/WriteHelpers.h [hunk 2]
```diff
@@ -581,6 +581,59 @@ void writeCSVString(const StringRef & s, WriteBuffer & buf)
     writeCSVString<quote>(s.data, s.data + s.size, buf);
 }
 
+inline void writeHTMLString(const char * begin, const char * end, WriteBuffer & buf)
```
Comment (alexey-milovidov) [inline src/IO/WriteHelpers.h:584]: The name is kind-of misleading.
The function is equally applicable to XML and HTML,
but unlike the `writeXMLString`, it also allows to place data into attributes.

We need to invent some better name for it...

File: src/IO/WriteHelpers.h (hunk 1 above)
Comment (nautaa) [reply src/IO/WriteHelpers.h:589]: In XML there are only five predefined character entity references, so I think it is not necessary.

File: src/IO/WriteHelpers.h (hunk 2 above)
Comment (nautaa) [reply src/IO/WriteHelpers.h:584]: I am Sorry. Would it be better to use writeXMLOrHTMLString?

Comment (nautaa) [reply src/IO/WriteHelpers.h:584]: I am not very clear about the use case.

Comment (alexey-milovidov) [reply src/IO/WriteHelpers.h:584]: Let's rename the old functon to `writeXMLStringForTextElement` and new function to `writeXMLStringForTextElementOrAttributeValue`.

Then we need to implement SQL function to use it.

Comment (nautaa) [reply src/IO/WriteHelpers.h:584]: thank you for your advice.

**Comment by nautaa** [comment]
@alexey-milovidov PTAL.

**Comment by alexey-milovidov** [comment]
Now we need to make an SQL function:

```
SELECT encodeXMLComponent('Hello, "world"!')
```

**Comment by nautaa** [comment]
@alexey-milovidov PTAL.

**Comment by nautaa** [comment]
I can pass the test locally, can you help me? @alexey-milovidov

**Comment by alexey-milovidov** [comment]
There is UBSan report: https://clickhouse-test-reports.s3.yandex.net/17659/00dc1040253259de6a77533fee5a05794e8fc8a6/functional_stateless_tests_(ubsan)/stderr.log

**Review by alexey-milovidov — APPROVED** [review]
Looks ready...

**Comment by alexey-milovidov** [comment]
@nautaa Congrats!
Now we need `decodeXMLComponent` function :)

**Comment by alexey-milovidov** [comment]
I also added performance test in follow-up PR.

**Comment by nautaa** [comment]
> @nautaa Congrats!
> Now we need `decodeXMLComponent` function :)

Thank you! I will try it later.

**Pull request merged**